### PR TITLE
Cache

### PR DESCRIPTION
**Pull request by artemnovichkov**
Update README with the cache logic description. Related to #14. BTW, add missing texts for arrays.

## Implement text style property observing

Implemented `Observer` pattern for text style object with flex memory management

### CODE REVIEW

**Review by stason314 — APPROVED** [review]

**Review by ntunin — APPROVED** [review]

**Review by ReshetovE — APPROVED** [review]

File: Texstyle/Texstyle/ControlStateText.swift [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    /// Initialize the text with passed string and styles for appropriate states. Returns nil if there is no value.
+    ///
+    /// - Parameters:
+    ///   - value: The string for style and substyles.
+    ///   - styles: The dictionary with states and text styles.
+    public convenience init?(value: String?, styles: [ControlState: TextStyle]) {
+        if let value = value {
+           self.init(value: value, styles: styles)
+           observeStyles()
```
Comment (stason314) [inline Texstyle/Texstyle/ControlStateText.swift]: `observeStyles()` was already called in `init(value: value, styles: styles)`. Should you call it here?

File: Texstyle/Texstyle/ControlStateText.swift [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                             for state: ControlState = .normal) -> CGRect {
+        guard let attributedString = attributed(for: state) else {
+           return .zero
+        }
+        return attributedString.boundingRect(with: size, options: options, context: context)
+    }
+
+    public func copy() -> ControlStateText {
+        let copy = ControlStateText(value: value, styles: styles.copy())
+        substyles.forEach { (state, substyles) in
```
Comment (stason314) [inline Texstyle/Texstyle/ControlStateText.swift]:
```suggestion
        substyles.forEach { state, substyles in
```

File: Texstyle/Texstyle/ControlStateText.swift [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+extension ControlStateText: TextStyleObserver {
+
+    func textStyleDidChange(_ textStyle: TextStyle) {
+        cachedAttributedStrings = [:]
+    }
+
+    private func observeStyles() {
+        let observerCenter = TextStyleObserverCenter.shared
+        observerCenter.remove(self)
+        styles.forEach { (_, style) in
```
Comment (stason314) [inline Texstyle/Texstyle/ControlStateText.swift]:
```suggestion
        styles.forEach { _, style in
```

File: Texstyle/Texstyle/ControlStateText.swift [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            observerCenter.add(self, for: style)
+        }
+        substyles.forEach { (_, substyles) in
```
Comment (stason314) [inline Texstyle/Texstyle/ControlStateText.swift]:
```suggestion
        substyles.forEach { _, substyles in
```

File: Texstyle/Texstyle/Text.swift [hunk 5]
```diff
@@ -8,87 +8,80 @@ import UIKit
...
+    /// Initialize the text with passed string and style. Returns nil if there is no value.
     ///
     /// - Parameters:
     ///   - value: The string for style and substyles.
-    ///   - substyle: The style for passed string.
+    ///   - style: The style for passed string.
     public convenience init?(value: String?, style: TextStyle) {
         if let value = value {
             self.init(value: value, style: style)
+            observeStyles()
```
Comment (stason314) [inline Texstyle/Texstyle/Text.swift]: Should you call it there ?

**Review by stason314 — CHANGES_REQUESTED** [review]

**Review by ReshetovE — APPROVED** [review]

**Review by stason314 — APPROVED** [review]

**Review by stason314 — APPROVED** [review]

**Comment by Antowkos** [comment]
@zhvrnkov, please add your review here

**Comment by ntunin** [comment]
@disabled Can you provide your feedback?

**Comment by ntunin** [comment]
@Antowkos @disabled

**Pull request merged**